### ISSUE TITLE
New RegexpSinglelineJava property ignoreStrings

### ISSUE THREAD

**Issue by robertwhitebit** [checkstyle/checkstyle#1405]
```
<module name="RegexpSinglelineJava">
  <property name="ignoreStrings" value="true" />
  ...
</module>
```
Strings should also be ignorable (like comments already are), because they can contain any valid text and therefore are likely to give false positive results.

**Comment by romani** [comment]
please provide example of real use case

**Comment by robertwhitebit** [comment]
Sure! I'll just use my current problem. :wink:

```
<module name="RegexpSinglelineJava">
  <property name="format" value="[^\s] {2,}" />
  <property name="ignoreComments" value="true" />
  <property name="ignoreStrings" value="true" />
  <property name="message" value="Multiple spaces for separation" />
</module>
```

This check prevents from separating tokens with more than one space, e.g.
```java
private static boolean   checkMergeConflicts(Project project) {
                       ↑
```

But falsely detects strings with multiple whitespaces, e.g.

```java
msg += "    " + spriteA.getName() + "\n";
         ↑
```

**Comment by romani** [comment]
Here is how suppression my comment is done:
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/CommentSuppressor.java

to do the same for constant, we will need to traverse tree by line number to catch that AST token and investigate all other tokens in case match is happen in first String value and then in code 

```java
msg += "    " + spriteA.getName()    + "\n";
         ↑                        ↑
```

that will be a lot of logic and processing to make such decision reliably. It is much better to create new Check for sequential space separation.

please provide more use cases if any, for now I do not see a reason to extend a RegExp Check.

**Comment by romani** [comment]
All RegExp Checks exists to let user be "crazy" with custom validations up to their own responsibility for false-positives. But as soon as validation become a bit more complicated, new Check need to be created.

**Comment by robertwhitebit** [comment]
Thanks, I took a quick look at the CommentSuppressor relevant code.

Theoretically there shouldn't be a (big) difference between multi-line comments and strings.
Both have a start and end tag and both can be multi-lined. The content of each is a valid text. The only catch is the escaped double quote ```\"```

> please provide more use cases if any, for now I do not see a reason to extend a RegExp Check.

Unfortunately I can't. As I can't for the ignoreComments property. My guess is, that many reasons which validates the ignoreComments property can be as well used for the ignoreStrings property. Maybe someone can provide other examples?

> ...new Check need to be created

Maybe the previous posted check is worth it's own checkstyle whitespace check?
Java tokens, which should separated by a whitespace, should always be separated by a single space, except you're using some special kind of horizontal aligning[1]

[1]
- Clean Code #Horizontal Alignment (discourages horizontal aligning)
- http://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment

> Tip: Alignment can aid readability, but it creates problems for future maintenance. Consider a future change that needs to touch just one line. This change may leave the formerly-pleasing formatting mangled, and that is allowed. More often it prompts the coder (perhaps you) to adjust whitespace on nearby lines as well, possibly triggering a cascading series of reformattings. That one-line change now has a "blast radius." This can at worst result in pointless busywork, but at best it still corrupts version history information, slows down reviewers and exacerbates merge conflicts.

**Comment by mkordas** [comment]
@robertwhitebit - I had exactly the same idea when I was working on #728. There is already sevntu-checkstyle/sevntu.checkstyle#334 issue created for such check - feel free to help us implement it (many people around the world including me will use it for sure).

**Comment by romani** [comment]
@robertwhitebit , I am voting against extending a RegExp Check, as it is way to making it smart to extend there it will be never reliable (or too much efforts to do validation that should be done in different approach). On implementing that users will report issues about false positives. We will not do implementation that is not reliable to reasonable extend.

One more workaround for you is extend that Check and use it as custom Check, there are hundreds custom Checks in that world, they are living ok out of main checkstyle project. Big companies have their own set of custom rules. That will fix your problem, and will let you control your code or detect all strange code.

Example (5 year old project): https://github.com/sevntu-checkstyle/sevntu.checkstyle 
Example of usage: https://github.com/sevntu-checkstyle/checkstyle-samples/blob/master/maven-project/pom.xml#L33 

I am closing that issue, but discussion could be continued if new ideas appeared.

**Comment by robertwhitebit** [comment]
Thanks for your feedback and guiding me to the [sevntu-checkstyle project](https://github.com/sevntu-checkstyle/sevntu.checkstyle), guys! :smiley_cat:

**Comment by solomax** [comment]
Hello @romani,

java17 introduces `Text blocks` maybe it worth to add property to ignore them? :)

**Comment by romani** [comment]
Please create new issue to discuss in detail.

**Comment by solomax** [comment]
Done! https://github.com/checkstyle/checkstyle/issues/11653
Thanks for the answer!
